### PR TITLE
Create the Boa v0.17 release post

### PR DESCRIPTION
**Pull request by Razican**
While this is still a very work-in-progress post, and we're still missing some features from reaching the `main` branch in Boa, this will be a HUGE release, so it would be great if we could already start writing the post.

If the post becomes too large, we might need to split it a bit and create multiple posts explaining some features (such as optimizations), and release them before the final release post, so that we can link back to them.

Things to do before the release:
 - [ ] Finish this post
 - [ ] ~New website using Docusaurus~
 - [ ] Clean-up issue tracker
    - [ ] Remove issues that have been fixed
    - [ ] Add newcommer / difficulty labels where missing
    - [ ] Prepare 0.18 milestone (should be much, much shorter than 0.17, 2 months max)
 - [ ] Merge all possible PRs, review the ones that can't be merged and update / close them

### CODE REVIEW

File: posts/2023-XX-YY-boa-release-17.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+x;
+"#;
+let js = Source::from_bytes(js_str);
+
+let mut context = Context::default();
+let res = context.eval(js).unwrap();
+
+let str = TestStruct::try_from_js(&res, &mut context)
+    .unwrap();
+
```
Comment (Razican) [inline posts/2023-XX-YY-boa-release-17.md]: I took this example from the examples directory, but it seems too long. Maybe we should reduce it, or just link it, what do you think?

File: posts/2023-XX-YY-boa-release-17.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+[ctx_provider]: https://boajs.dev/boa/doc/boa_engine/context/struct.ContextBuilder.html#method.icu_provider
+[data]: https://github.com/unicode-org/icu4x/blob/main/docs/tutorials/data_management.md#data-management-in-icu4x
+[`icu4x`]: https://github.com/unicode-org/icu4x
+
+#### `WeakRef`, `WeakSet` and `WeakMap`
+
+We've implemented support for weak references to garbage collected objects. This allowed us to
+implement some builtins like [`WeakRef`][wr], [`WeakSet`][ws] and [`WeakMap`][wm]. Note that the
+garbage collection algorithm is always subject to future optimizations and modifications, so it is
+best to avoid using them if possible.
```
Comment (Razican) [inline posts/2023-XX-YY-boa-release-17.md]: This seems to imply that `WeakRef`, `WeakSet` and `WeakMap` shouldn't be used by JavaScript users. Is this the case?

Comment (jedel1043) [reply posts/2023-XX-YY-boa-release-17.md]: Yeah, that is the suggestion from [Mozilla's docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef#avoid_where_possible)

> Correct use of WeakRef takes careful thought, and it's best avoided if possible.

File: posts/2023-XX-YY-boa-release-17.md (hunk 1 above)
Comment (HalidOdat) [reply posts/2023-XX-YY-boa-release-17.md]: I slightly reduce TryFromJs example in the latest commit, IMO it's not too big.

File: posts/2023-XX-YY-boa-release-17.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+#### Object Shapes (Hidden classes)
+
+Hidden classes (in boa called "shapes" to avoid confusion with JavaScript classes) is an optimization technique that aims to improve the performance of dynamically-typed programming languages. This optimization works by analyzing the patterns of object creation and dynamically generating hidden classes to represent similar object structures.
+
+In dynamically-typed languages, objects are usually a bag of properties (key and value pairs) that are modified at runtime, which can lead to inefficiencies in terms of memory usage. Hidden classes optimization addresses this by organizing object structures more efficiently, reducing memory overhead, it also unlocks the posability for other optimizations, such as inline caching, which optimizes property access.
```
Comment (jedel1043) [inline posts/2023-XX-YY-boa-release-17.md]:
```suggestion
In dynamically-typed languages, objects are usually a bag of properties (key and value pairs) that are modified at runtime, which can lead to inefficiencies in terms of memory usage. Hidden classes optimization addresses this by organizing object structures more efficiently, reducing memory overhead, it also unlocks the possibility for other optimizations, such as inline caching, which optimizes property access.
```

Also, I think this should first explain what exactly is hidden classes, then explain the upsides of it:
>  Hidden classes are an alternative way to structure objects that \<how hidden classes work\>. With this, objects are stored more efficiently, \<rest of the upsides\>.

File: posts/2023-XX-YY-boa-release-17.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+The shapes create a trasition tree, where the trasitions are property and prototype changes starting from the root shape (no properties, no prototype).
```
Comment (nekevss) [inline posts/2023-XX-YY-boa-release-17.md]:
```suggestion
The shapes create a transition tree, where the transitions are property and prototype changes starting from the root shape (no properties, no prototype).
```

File: posts/2023-XX-YY-boa-release-17.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+When creating objects with the same property names, it's best to create them in the same order, this ensures that the objects share the same shape.
+
+For a more in depth explanation of how shared shapes work in boa see `shapes.md` [here](https://github.com/boa-dev/boa/blob/main/docs/shapes.md).
+
+TO-DO: explain the enhancements of new optimizations (constant folding, inline catching, hidden classes...)
+
+### Debug object
+
+If you're debugging your JavaScript code with the Boa CLI interface, you can use the new `$boa` object. If you want to use it,
```
Comment (nekevss) [inline posts/2023-XX-YY-boa-release-17.md]: This should probably begin a bit more assertive.
```suggestion
The `$boa` debug object has been implemented for convenient JavaScript debugging using Boa's CLI interface. If you want to use it,
```

File: posts/2023-XX-YY-boa-release-17.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+The `$boa` debug object is divided into modules, so that you can trigger the garbage collection with `$boa.gc.collect()`, or
+get the bytecode of a function by running `$boa.function.bytecode(fn_name)`. You can also trace function invocations, handle
+compiler optimizations, limits and object shapes.
+
+You can find all the documentation [here][debug_object].
+
+[debug_object]: https://github.com/boa-dev/boa/blob/main/docs/boa_object.md
+
+### Modules
```
Comment (nekevss) [inline posts/2023-XX-YY-boa-release-17.md]: Should the `Modules` section be moved farther up in the post? It's a pretty big update.

File: posts/2023-XX-YY-boa-release-17.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+- [`Intl.ListFormat`][list]
+- [`Intl.Locale`][locale]
+- [`Intl.Segmenter`][segmenter]
+
+However, internationalization data can be pretty expensive at times; the default data included by Boa
+is 10.6 MB in size; which is why we allow customizing the data provider used by the engine with the
+[`ContextBuilder::icu_provider`][ctx_provider] hook. For more information on how to generate custom
+internationalization data, you can check out the [**data management tutorial**][data] from [`icu4x`],
+the internationalization library used in Boa. Shoutout to the `icu4x` team, which are the ones that
+made all of this possible!
```
Comment (nekevss) [inline posts/2023-XX-YY-boa-release-17.md]: Small suggestion on the wording here. 😄 
```suggestion
Internationalization data can be pretty expensive at times: the default data included by Boa
is 10.6 MB, which is why we allow customizing the data provider used by the engine with the
[`ContextBuilder::icu_provider`][ctx_provider] hook. For more information on how to generate custom
internationalization data, you can check out the [**data management tutorial**][data] from [`icu4x`],
the internationalization library used in Boa. Shoutout to the `icu4x` team, who are the ones that
made all of this possible!
```

File: posts/2023-XX-YY-boa-release-17.md (hunk 2 above)
Comment (Razican) [reply posts/2023-XX-YY-boa-release-17.md]: I would reword this, since the current explanation seems to imply that the reason is potential future incompatibility with some optimizations, but Mozilla seems to imply that this is due to the complexity associated with using these types.

File: posts/2023-XX-YY-boa-release-17.md [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+There have also been a various number of other internal enhancements made.
+
+#### Split Node into Statement / Declaration / Expression
+
+In the last release, Boa's AST used a `Node` enum to represent both the `Statement`, `Declaration` and `Expression` parse nodes. One of the large internal improvements made for this release was to split `Node` into `Statement`, `Declaration`, and `Expression` nodes. This refactor involved not only large changes to the AST but also further changes to the bytecompiler and parser. The split also brings us closer in line with the ECMAScript specification.
+
+#### UTF-16 strings
+
+With this release, Boa's `JsString`s are now implemented as utf-16 encoded strings. Along with the new `JsString`, there are two provided macros: `js_string!` for creating a new `JsString` from a `&str`, and `utf16` for creating a utf-16 array literal from a `&str`.
```
Comment (RageKnify) [inline posts/2023-XX-YY-boa-release-17.md]:
```suggestion
With this release, Boa's `JsString`s are now implemented as utf-16 encoded strings. Along with the new `JsString`, there are two provided macros: `js_string!` for creating a new `JsString` from a `&str`, and `utf16!` for creating a utf-16 array literal from a `&str`.
```
This refers to a macro, right?

File: posts/2023-XX-YY-boa-release-17.md [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+let module = Module::parse(source, None, context).unwrap();
+
+loader.insert(Path::new("main.mjs").to_path_buf(), module.clone());
+
+let promise = module.load_link_evaluate(context).unwrap();
+
+context.run_jobs();
+
+assert_eq!(promise.state().unwrap(), PromiseState::Fulfilled(JsValue::undefined()));
+```
```
Comment (jasonwilliams) [inline posts/2023-XX-YY-boa-release-17.md]: I think this will need some comments

**Comment by jasonwilliams** [comment]
The changes I've just pushed are the following:
- adds default build
- improves the spacing for wider screens
- Defers loading of mermaid.js (most mermaid.js usage is below the fold)

The deferred loading of mermaid speeds up the site by around 150ms

File: posts/2023-07-05-boa-release-17.md [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+
+assert_eq!(&msg, utf16!("Hello, human! Nice to meet you!"));
+```
+
+## Conclusions
+
+If you reached so far, you probably understand how big this release was, and you can find even more changes in the
+[full changelog][changelog]. Boa is now becoming a real option for many projects, which shows with the amount of
+financial support we have received these last months. Nevertheless, going forward, we need your help to get to a 1.0
+version. Wheter you are good with Rust, JavaScript, documentation or development, we have multiple
```
Comment (jedel1043) [inline posts/2023-07-05-boa-release-17.md]:
```suggestion
version. Whether you are good with Rust, JavaScript, documentation or development, we have multiple
```

File: posts/2023-07-05-boa-release-17.md [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+[good first issues][first_issues], and places where we [need help][help_wanted], both in Boa's
+[main repository][boa_engine] and [others][boa_org] around it.
+
+Once again, we have to thank to [all the contributors][contributors] of this release!!
```
Comment (jedel1043) [inline posts/2023-07-05-boa-release-17.md]:
```suggestion
Once again, big thanks to [all the contributors][contributors] of this release!!
```

**Review by jedel1043 — APPROVED** [review]

**Review by HalidOdat — APPROVED** [review]

File: posts/2023-07-08-boa-release-17.md [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+
+In this release, our conformance has grown from 74.53% to 78.74% in the official ECMAScript Test Suite (Test262). While
+this might look as a small increase, we now pass **6,079** more tests than in the previous version. In any case, the big
+changes in this release are not related to conformance, but to huge internal enhancements and new APIs that you will be
+able to use.
+
+You can check the full list of changes [here][changelog], and the full information on conformance [here][conformance].
+
+Moreover, this big release was partly possible thanks to a grant by [Lit Protocol](https://litprotocol.com/). Lit is
+distributed key management for encryption, signing, and compute. Thanks to this grant, we were able to remunerate 2 team
```
Comment (jedel1043) [inline posts/2023-07-08-boa-release-17.md]: I think we can omit explaining what lit protocol is, the hyperlink takes care of that for us.

File: posts/2023-07-08-boa-release-17.md [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+context.run_jobs();
+
+// All modules return `undefined` if they're successfully evaluated.
+assert_eq!(promise.state().unwrap(), PromiseState::Fulfilled(JsValue::undefined()));
+```
+
+For a more extensive, descriptive example that uses a real directory, you can check out [**boa_examples**][mod_example].
+
+[modules]: https://tc39.es/ecma262/#sec-modules
+[mod_loader]: https://boajs.dev/boa/doc/boa_engine/module/trait.ModuleLoader.html
```
Comment (Razican) [inline posts/2023-07-08-boa-release-17.md]: We should link to the 0.17 version of the docs
```suggestion
[mod_loader]: https://docs.rs/boa_engine/0.17.0/boa_engine/module/trait.ModuleLoader.html
```

File: posts/2023-07-08-boa-release-17.md [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+[mod_example]: https://github.com/boa-dev/boa/blob/main/boa_examples/src/bin/modules.rs
```
Comment (Razican) [inline posts/2023-07-08-boa-release-17.md]:
```suggestion
[mod_example]: https://github.com/boa-dev/boa/blob/v0.17/boa_examples/src/bin/modules.rs
```

File: posts/2023-07-08-boa-release-17.md [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+
+### Spec Version Conformance
+
+Something we get asked a lot is _"Do you support ES5 or ES6"_? or _"How far away are you from supporting ESX"_?
+We're pleased to say we've updated our conformance board to show you how we're doing across ES versions.
+
+Just navigate to our [Test262 Dashboard](https://boajs.dev/boa/test262/), select "Test Results" on our main branch, and
```
Comment (Razican) [inline posts/2023-07-08-boa-release-17.md]:
```suggestion
Just navigate to our [Test262 Dashboard][conformance], select "Test Results" on our main branch, and
```

File: posts/2023-07-08-boa-release-17.md [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+```
+
+This separation of property keys and values allows for objects with the same property names to share the same shape,
+which reduces memory consumption and unlocks the possibility for other optimizations such as inline caching.
+
+Note: When creating objects with the same property keys, it's best to create them in the same order, this ensures that
+the objects share the same shape.
+
+For a more in depth explanation of how shared shapes work in boa see `shapes.md`
+[here](https://github.com/boa-dev/boa/blob/main/docs/shapes.md).
```
Comment (Razican) [inline posts/2023-07-08-boa-release-17.md]:
```suggestion
[here](https://github.com/boa-dev/boa/blob/v0.17/docs/shapes.md).
```

File: posts/2023-07-08-boa-release-17.md [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+The `$boa` debug object has been implemented for convenient JavaScript debugging using Boa's CLI interface. If you want
+to use it, you will need to run the Boa CLI / REPL with the `--debug-object` command line flag.
+
+The `$boa` debug object is divided into modules, so that you can trigger the garbage collection with `$boa.gc.collect()`,
+or get the bytecode of a function by running `$boa.function.bytecode(fn_name)`. You can also trace function invocations,
+handle compiler optimizations, set runtime limits and inspect object shapes.
+
+You can find all the documentation [here][debug_object].
+
+[debug_object]: https://github.com/boa-dev/boa/blob/main/docs/boa_object.md
```
Comment (Razican) [inline posts/2023-07-08-boa-release-17.md]:
```suggestion
[debug_object]: https://github.com/boa-dev/boa/blob/v0.17/docs/boa_object.md
```

File: posts/2023-07-08-boa-release-17.md [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+use boa_engine::{Context, Source};
+
+fn main() {
+    let js_file_path = "./scripts/helloworld.js";
+
+    match Source::from_filepath(Path::new(js_file_path)) {
+    ...
+```
+
+See Boa's [examples](https://github.com/boa-dev/boa/tree/main/boa_examples/src/bin) for more examples on how its used.
```
Comment (Razican) [inline posts/2023-07-08-boa-release-17.md]:
```suggestion
See Boa's [examples](https://github.com/boa-dev/boa/tree/v0.17/boa_examples/src/bin) for more examples on how its used.
```

File: posts/2023-07-08-boa-release-17.md [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+this might look as a small increase, we now pass **6,079** more tests than in the previous version. In any case, the big
+changes in this release are not related to conformance, but to huge internal enhancements and new APIs that you will be
+able to use.
+
+You can check the full list of changes [here][changelog], and the full information on conformance [here][conformance].
+
+Moreover, this big release was partly possible thanks to a grant by [Lit Protocol](https://litprotocol.com/). Lit is
+distributed key management for encryption, signing, and compute. Thanks to this grant, we were able to remunerate 2 team
+members for their 20h/week work each during three and a half months. If you wish to sponsor Boa, you can do so by
+donating to out [open collective][collective]. You can also check [easy][easy_issues] or
```
Comment (RageKnify) [inline posts/2023-07-08-boa-release-17.md]: *donating to **our**

File: posts/2023-07-08-boa-release-17.md (hunk 12 above)
Comment (Razican) [reply posts/2023-07-08-boa-release-17.md]: Done :)

File: posts/2023-07-08-boa-release-17.md (hunk 19 above)
Comment (Razican) [reply posts/2023-07-08-boa-release-17.md]: Fixed

**Review by jedel1043 — APPROVED** [review]
